pkg/daemon/ceph/client/subvolumegroup.go

```diff
@@ -55,3 +58,71 @@ func DeleteCephFSSubVolumeGroup(context *clusterd.Context, clusterInfo *ClusterI
 	logger.Infof("successfully deleted cephfs %q subvolume group %q", volName, groupName)
 	return nil
 }
+
+// PinCephFSSubVolumeGroup pin the cephfs subvolume group
+func PinCephFSSubVolumeGroup(context *clusterd.Context, clusterInfo *ClusterInfo, volName string, cephFilesystemSubVolumeGroup *cephv1.CephFilesystemSubVolumeGroup, cephFilesystemSubVolumeGroupName string) error {
+	logger.Infof("pinning cephfs subvolume group %q of filesystem %q", cephFilesystemSubVolumeGroupName, volName)
+
+	args := []string{"fs", "subvolumegroup", "pin", volName, cephFilesystemSubVolumeGroupName}
+	if cephFilesystemSubVolumeGroup.Spec.Pinning.Distributed != nil {
+		setting := strconv.Itoa(*cephFilesystemSubVolumeGroup.Spec.Pinning.Distributed)
+		args = append(args, "distributed", setting)
+	} else if cephFilesystemSubVolumeGroup.Spec.Pinning.Export != nil {
+		setting := strconv.Itoa(*cephFilesystemSubVolumeGroup.Spec.Pinning.Export)
+		args = append(args, "export", setting)
+	} else if cephFilesystemSubVolumeGroup.Spec.Pinning.Random != nil {
+		setting := strconv.FormatFloat(*cephFilesystemSubVolumeGroup.Spec.Pinning.Random, 'f', -1, 64)
+		args = append(args, "random", setting)
+	} else {
+		// set by default value
+		args = append(args, "distributed", "1")
+	}
+	logger.Infof("subvolume group pinning args %v", args)
+
+	err := validatePinningValues(cephFilesystemSubVolumeGroup.Spec.Pinning)
```

Comment:
Generally, I expect to find validation code before code that processes values. This is pretty minor in this case because validation is still done before the command is executed. It would be good to move it so that it matches trends that other developers expect. That helps the code be more well understood by others, and it helps the code keep correctness if someone needs to make changes in the future that might change the nature of the preceding code.

Reply:
Updated